pkg/utils/fswatcher.go

```diff
@@ -130,12 +148,24 @@ func (w *FsWatcher) watch() {
 
 			// If any of our paths change
 			name := filepath.Clean(e.Name)
-			if _, ok := w.paths[filepath.Clean(name)]; ok {
-				klog.V(2).Infof("fsnotify %s event in %q detected", e, name)
+			if _, ok := w.paths[name]; ok {
+				klog.Infof("fsnotify %s event in %q detected", e, name)
+				// Creation and deletion events should always trigger reloads since stat may fail
+				if e.Op&fsnotify.Create == fsnotify.Create || e.Op&fsnotify.Remove == fsnotify.Remove {
+					ratelimiter = time.After(w.ratelimit)
+					continue
+				}
+				// Did the file really change according to the modtimes?
+				statResult, err := os.Stat(name)
+				if err != nil {
+					klog.Errorf("Failed to check modtime on file %s: %v", name, err)
+					return
+				}
+				if statResult.ModTime() != w.modtimes[name] {
+					w.modtimes[name] = statResult.ModTime()
+					ratelimiter = time.After(w.ratelimit)
+				}
```

Comment:
I'd suggest to change it the other way around. IFF `op == fsnotify.Write` then check mtime. Othewise (all other ops) always reconfigure

Reply:
If all operations other than `fsnotify.Write` were to trigger reconfiguration, that would still have the bug in #805 because `fsnotify.Chmod` would trigger reloads unnecessarily in k8s.

For context, these are the events when the file contents don't change:
```
"features.d": CHMOD
"source.d": CHMOD
"nfd-worker.conf": CHMOD
"..data": CHMOD
"nfd-worker.conf": CHMOD
"..2022_04_14_14_40_56.720262905": CHMOD
"./": CHMOD
```

And these are the events when the file does change (note there aren't any write events. The file that gets written is inside the `..2022_...` folder which we aren't watching directly)

```
"..2022_04_26_13_44_40.959076891": CREATE
"..2022_04_26_13_44_40.959076891": CHMOD
"..data_tmp": RENAME
"..data": CREATE
"nfd-worker.conf": CHMOD
"nfd-worker.conf": REMOVE
"..2022_04_14_14_40_56.720262905": REMOVE
"features.d": CHMOD
"source.d": CHMOD
"nfd-worker.conf": CHMOD
"..data": CHMOD
"..2022_04_26_13_44_40.959076891": CHMOD
"./": CHMOD
```

Reply:
This is hairy :/ You cannot rely on mtime if chmod occurs. Think about `chmod -r nfdworker.conf`

Reply:
Chmod doesn't change mtime:

```
/tmp # touch myfile.txt
/tmp # touch myfile2.txt
/tmp # ls -alh
total 0
drwxrwxrwt    1 root     root          43 Apr 26 14:44 .
drwxr-xr-x    1 root     root          40 Apr 26 14:40 ..
-rw-r--r--    1 root     root           0 Apr 26 14:44 myfile.txt
-rw-r--r--    1 root     root           0 Apr 26 14:44 myfile2.txt

# wait 1 minute

/tmp # chmod 777 myfile2.txt
/tmp # ls -alh
total 0
drwxrwxrwt    1 root     root          43 Apr 26 14:44 .
drwxr-xr-x    1 root     root          40 Apr 26 14:40 ..
-rw-r--r--    1 root     root           0 Apr 26 14:44 myfile.txt
-rwxrwxrwx    1 root     root           0 Apr 26 14:44 myfile2.txt
```

Reply:
> Chmod doesn't change mtime:

Exactly. But it could make the file unreadable/readable. The watcher implementation was implemented probably stupidly bombproof but I wouldn't want to break it now. For example, it watches the whole directory chain (up to the root dir) and this PR does not seen ti take that into account. Looking at this now, I would actually keep the `FsWatcher` stupid and simple as it is now (generic watcher for all kinds of "files", including directories, devices, sockets...) and implement a new `ConfigFileWatcher` or smth on top of that

With this in mind it looks, again, that we should probably just resort to comparison/checksumming. The amount of data is not expected to be huge. An alternative approach could be to cache the mtime and readability status of the file (trying to think what kind of caveats that approach contains but cannot think of any atm 😄).

Reply:
> The watcher implementation was implemented probably stupidly bombproof but I wouldn't want to break it now.

No behavior is being broken; this implementation and the old one does the same thing: Ignores the change until the file is readable again.

You can test by changing this `os.Remove` to an `os.Chmod(configFile, 0100)`: https://github.com/kubernetes-sigs/node-feature-discovery/blob/041364dfc6c06b8204177c9a92c1afe2c6600072/pkg/nfd-client/worker/nfd-worker-internal_test.go#L254

The old code fails the test because it just logs a warning and doesn't trigger a reload: https://github.com/kubernetes-sigs/node-feature-discovery/blob/041364dfc6c06b8204177c9a92c1afe2c6600072/pkg/utils/fswatcher.go#L141-L147

The new code also fails the test because a chmod just doesn't trigger a reload.

Besides, I don't think we should be worrying too much about `chmod`. You can't even run `chmod` inside the container:

```
$ docker run -it gcr.io/k8s-staging-nfd/node-feature-discovery:master sh -c 'chmod 600 /etc/kubernetes/node-feature-discovery/nfd-worker.conf'
chmod: changing permissions of '/etc/kubernetes/node-feature-discovery/nfd-worker.conf': Operation not permitted
```